File: src/sooljang/domain/metrics.py
"""파생 지표 계산.

`docs/architecture.md` §3 의 수식을 순수 함수로 구현한다. SQLAlchemy·HTTP 를 import 하지
않으므로 DB 없이 단위 테스트할 수 있다. 같은 공식을 SQL 로도 구현하고
(`infrastructure/database/metrics_sql.py`) 두 결과가 일치하는지 테스트로 보장한다.

**파생값은 저장하지 않는다.** 저장하면 엑셀과 같은 불일치 문제가 재발한다(§9.5).

핵심 규칙 세 가지를 기억할 것.

1. **100ml당 가격은 정가 기준**이다. 레거시 통계와 일치시키기 위한 결정이며
   실구매 기준으로 계산하면 168건이 불일치한다(`docs/legacy-schema.md` §4.2)
2. **가격이 없는 구매 건(선물)은 금액 집계에서 제외**하되 병수 집계에는 포함한다.
   레거시에 정가 결측 33건, 실구매가 결측 34건이 있다
3. **여러 용량이 섞인 제품**의 100ml당 가격은 가중 평균으로 계산한다
"""

import datetime
from collections.abc import Iterable, Sequence
from dataclasses import dataclass, field
from decimal import ROUND_HALF_UP, Decimal

#: 금액을 원 단위 소수 둘째 자리까지 유지한다. 표시 단계에서 반올림한다.
MONEY_QUANT = Decimal("0.01")
#: 비율(할인율 등)은 소수 넷째 자리까지 유지한다.
RATIO_QUANT = Decimal("0.0001")
#: 100ml 단위 환산 계수.
ML_PER_UNIT = Decimal(100)


def quantize_money(value: Decimal) -> Decimal:
    return value.quantize(MONEY_QUANT, rounding=ROUND_HALF_UP)


def quantize_ratio(value: Decimal) -> Decimal:
    return value.quantize(RATIO_QUANT, rounding=ROUND_HALF_UP)


class BottleState:
    """병 상태 문자열. 도메인 계층은 ORM enum 에 의존하지 않는다.

    도메인이 `infrastructure` 를 import 하면 의존 방향이 뒤집힌다. 값은 ORM 의
    `BottleStatus` 와 같게 유지하고, 어긋나면 테스트가 잡는다.
    """

    UNOPENED = "unopened"
    OPEN = "open"
    FINISHED = "finished"
    GIFTED = "gifted"
    SOLD = "sold"


IN_STOCK_STATES = frozenset({BottleState.UNOPENED, BottleState.OPEN})
DISPOSED_STATES = frozenset({BottleState.GIFTED, BottleState.SOLD})


@dataclass(frozen=True, slots=True)
class PurchaseLot:
    """구매 건 하나의 금액·수량과 그 규격의 용량."""

    quantity: int
    volume_ml: int
    unit_list_price: Decimal | None = None
    unit_paid_price: Decimal | None = None

    def __post_init__(self) -> None:
        if self.quantity <= 0:
            raise ValueError(f"구매 병수는 양수여야 합니다: {self.quantity}")
        if self.volume_ml <= 0:
            raise ValueError(f"용량은 양수여야 합니다: {self.volume_ml}")

    @property
    def list_total(self) -> Decimal | None:
        """정가 총액. 단가가 없으면 None."""
        if self.unit_list_price is None:
            return None
        return self.unit_list_price * Decimal(self.quantity)

    @property
    def paid_total(self) -> Decimal | None:
        """실지불 총액. 단가가 없으면 None."""
        if self.unit_paid_price is None:
            return None
        return self.unit_paid_price * Decimal(self.quantity)

    @property
    def total_volume_ml(self) -> int:
        return self.volume_ml * self.quantity


@dataclass(frozen=True, slots=True)
class BottleRecord:
    """병 하나의 상태와 날짜."""

    status: str
    opened_on: datetime.date | None = None
    finished_on: datetime.date | None = None

    @property
    def in_stock(self) -> bool:
        return self.status in IN_STOCK_STATES

    @property
    def days_to_finish(self) -> int | None:
        """개봉 후 소진까지 걸린 일수. 둘 중 하나라도 없으면 None."""
        if self.opened_on is None or self.finished_on is None:
            return None
        return (self.finished_on - self.opened_on).days


@dataclass(frozen=True, slots=True)
class BottleTally:
    """상태별 병수 집계."""

    unopened: int = 0
    open: int = 0
    finished: int = 0
    gifted: int = 0
    sold: int = 0

    @property
    def in_stock(self) -> int:
        """재고 병수. 증여·판매·소진은 재고가 아니다."""
        return self.unopened + self.open

    @property
    def disposed(self) -> int:
        return self.gifted + self.sold

    @property
    def total(self) -> int:
        return self.in_stock + self.finished + self.disposed


def tally_bottles(bottles: Iterable[BottleRecord]) -> BottleTally:
    """병 목록을 상태별로 집계한다. 알 수 없는 상태는 세지 않고 무시한다."""
    counts = dict.fromkeys(
        (
            BottleState.UNOPENED,
            BottleState.OPEN,
            BottleState.FINISHED,
            BottleState.GIFTED,
            BottleState.SOLD,
        ),
        0,
    )
    for bottle in bottles:
        if bottle.status in counts:
            counts[bottle.status] += 1
    return BottleTally(
        unopened=counts[BottleState.UNOPENED],
        open=counts[BottleState.OPEN],
        finished=counts[BottleState.FINISHED],
        gifted=counts[BottleState.GIFTED],
        sold=counts[BottleState.SOLD],
    )


@dataclass(frozen=True, slots=True)
class PriceMetrics:
    """금액 관련 파생 지표."""

    purchased_count: int
    priced_quantity: int
    paid_quantity: int
    list_total: Decimal | None
    paid_total: Decimal | None
    avg_list_price: Decimal | None
    avg_paid_price: Decimal | None
    price_per_100ml: Decimal | None
    price_per_100ml_paid: Decimal | None
    discount_rate: Decimal | None
    total_volume_ml: int

    @property
    def has_prices(self) -> bool:
        return self.avg_list_price is not None or self.avg_paid_price is not None


def _weighted_total(lots: Sequence[PurchaseLot], *, paid: bool) -> tuple[Decimal | None, int, int]:
    """가격이 있는 구매 건만 합산한다.

    반환값은 (금액 합계, 병수 합계, 용량 합계 ml). 금액이 있는 구매 건이 없으면 금액은
    None 이다. 0 을 반환하면 "전부 무료" 와 "가격 정보 없음" 을 구분할 수 없다.
    """
    amount = Decimal(0)
    quantity = 0
    volume = 0
    found = False
    for lot in lots:
        total = lot.paid_total if paid else lot.list_total
        if total is None:
            continue
        found = True
        amount += total
        quantity += lot.quantity
        volume += lot.total_volume_ml
    return (amount if found else None), quantity, volume


def compute_price_metrics(lots: Sequence[PurchaseLot]) -> PriceMetrics:
    """구매 건 목록에서 금액 파생 지표를 계산한다."""
    purchased = sum(lot.quantity for lot in lots)
    total_volume = sum(lot.total_volume_ml for lot in lots)

    list_total, list_quantity, list_volume = _weighted_total(lots, paid=False)
    paid_total, paid_quantity, paid_volume = _weighted_total(lots, paid=True)

    avg_list = (
        quantize_money(list_total / Decimal(list_quantity))
        if list_total is not None and list_quantity > 0
        else None
    )
    avg_paid = (
        quantize_money(paid_total / Decimal(paid_quantity))
        if paid_total is not None and paid_quantity > 0
        else None
    )

    # 여러 용량이 섞이면 가중 평균이어야 한다. 단일 용량이면 avg / (volume/100) 과 같다.
    per_100ml = (
        quantize_money(list_total * ML_PER_UNIT / Decimal(list_volume))
        if list_total is not None and list_volume > 0
        else None
    )
    per_100ml_paid = (
        quantize_money(paid_total * ML_PER_UNIT / Decimal(paid_volume))
        if paid_total is not None and paid_volume > 0
        else None
    )

    discount = _discount_rate(lots)

    return PriceMetrics(
        purchased_count=purchased,
        priced_quantity=list_quantity,
        paid_quantity=paid_quantity,
        list_total=list_total,
        paid_total=paid_total,
        avg_list_price=avg_list,
        avg_paid_price=avg_paid,
        price_per_100ml=per_100ml,
        price_per_100ml_paid=per_100ml_paid,
        discount_rate=discount,
        total_volume_ml=total_volume,
    )


def _discount_rate(lots: Sequence[PurchaseLot]) -> Decimal | None:
    """할인율. 정가와 실구매가가 **모두 있는** 구매 건만으로 계산한다.

    한쪽만 있는 구매 건을 섞으면 분모와 분자의 모집단이 달라져 할인율이 왜곡된다.
    """
    list_amount = Decimal(0)
    paid_amount = Decimal(0)
    found = False
    for lot in lots:
        if lot.unit_list_price is None or lot.unit_paid_price is None:
            continue
        found = True
        list_amount += lot.unit_list_price * Decimal(lot.quantity)
        paid_amount += lot.unit_paid_price * Decimal(lot.quantity)
    if not found or list_amount == 0:
        return None
    return quantize_ratio(Decimal(1) - paid_amount / list_amount)


@dataclass(frozen=True, slots=True)
class ProductMetrics:
    """제품 하나의 전체 파생 지표."""

    prices: PriceMetrics
    bottles: BottleTally
    inventory_value_at_cost: Decimal | None
    average_days_to_finish: Decimal | None
    consistency_warnings: tuple[str, ...] = field(default=())

    @property
    def in_stock_count(self) -> int:
        return self.bottles.in_stock

    @property
    def consumed_count(self) -> int:
        return self.bottles.finished


def compute_product_metrics(
    lots: Sequence[PurchaseLot], bottles: Sequence[BottleRecord]
) -> ProductMetrics:
    """제품 수준 파생 지표를 계산한다.

    병수 정합성이 어긋나면 예외 대신 경고를 남긴다. 레거시에서 넘어온 데이터가
    완벽하지 않을 수 있고, 지표를 아예 못 보는 것보다 경고와 함께 보는 것이 낫다.
    """
    prices = compute_price_metrics(lots)
    tally = tally_bottles(bottles)

    warnings: list[str] = []
    if tally.total != prices.purchased_count:
        warnings.append(
            f"병 레코드 수({tally.total})가 구매 병수({prices.purchased_count})와 다릅니다"
        )

    inventory_value = (
        quantize_money(prices.avg_paid_price * Decimal(tally.in_stock))
        if prices.avg_paid_price is not None
        else None
    )

    durations = [bottle.days_to_finish for bottle in bottles if bottle.days_to_finish is not None]
    average_days = (
        quantize_ratio(Decimal(sum(durations)) / Decimal(len(durations))) if durations else None
    )

    return ProductMetrics(
        prices=prices,
        bottles=tally,
        inventory_value_at_cost=inventory_value,
        average_days_to_finish=average_days,
        consistency_warnings=tuple(warnings),
    )


def value_for_money(rating: Decimal | None, price_per_100ml: Decimal | None) -> Decimal | None:
    """가성비. 100ml당 가격 1,000원 기준으로 정규화해 값이 너무 작아지지 않게 한다.

    높을수록 좋다. 가격이 0이면 정의되지 않는다.
    """
    if rating is None or price_per_100ml is None or price_per_100ml <= 0:
        return None
    return quantize_ratio(rating * Decimal(1000) / price_per_100ml)


def consumption_rate_per_month(finished_count: int, days: int) -> Decimal | None:
    """소비 속도(월 환산 병수). 기간이 0이면 정의되지 않는다."""
    if days <= 0:
        return None
    return quantize_ratio(Decimal(finished_count) * Decimal(30) / Decimal(days))


def krw_from_foreign(foreign_amount: Decimal | None, fx_rate: Decimal | None) -> Decimal | None:
    """외화 금액을 구매 시점 환율로 원화 환산한다.

    환율은 구매 시점 값으로 고정한다. 현재 환율로 재평가하면 과거 지출액이 매일 바뀌어
    통계가 재현되지 않는다(`docs/architecture.md` §2.3 `purchase`).
    """
    if foreign_amount is None or fx_rate is None:
        return None
    return quantize_money(foreign_amount * fx_rate)

File: web/src/domain/metrics.ts
/**
 * 파생 지표 계산 — `src/sooljang/domain/metrics.py` 의 TypeScript 포팅.
 *
 * 오프라인 상태에서도 제품 목록·통계 화면을 보여주려면(사용자가 선택한 "전체 컬렉션
 * 오프라인 탐색" 범위) Dexie 미러의 원자값으로 같은 공식을 다시 계산해야 한다. Dexie 는
 * 원자값만 저장하고 파생 지표는 저장하지 않는다 — 이 원칙은 서버와 동일하다.
 *
 * Python 쪽엔 이미 domain(순수 함수) vs SQL 의 2-way parity 테스트가 있다(Task 8). 여기
 * TS 구현이 3번째로 더해지므로, `tests/fixtures/metrics_cases.json` 공유 골든값을 통해
 * 세 구현이 갈라지지 않는지 확인한다(`web/src/domain/metrics.test.ts`,
 * `tests/domain/test_metrics.py::test_price_metrics_matches_shared_fixture`).
 *
 * **핵심 규칙 세 가지** (Python 원본과 동일):
 * 1. 100ml당 가격은 정가 기준이다
 * 2. 가격이 없는 구매 건(선물)은 금액 집계에서 제외하되 병수 집계에는 포함한다
 * 3. 여러 용량이 섞인 제품은 가중 평균으로 계산한다
 *
 * `consumption_rate_per_month`·`krw_from_foreign` 은 포팅하지 않았다 — 오프라인 화면
 * 어디에서도 쓰지 않는다. 필요해지면 그때 추가한다. `value_for_money` 는 Task 20 에서
 * 제품 지표에 노출하며 포팅했다.
 */

import Decimal from "decimal.js";

const MONEY_DECIMALS = 2;
const RATIO_DECIMALS = 4;
const ML_PER_UNIT = 100;

export function quantizeMoney(value: Decimal): Decimal {
  return value.toDecimalPlaces(MONEY_DECIMALS, Decimal.ROUND_HALF_UP);
}

export function quantizeRatio(value: Decimal): Decimal {
  return value.toDecimalPlaces(RATIO_DECIMALS, Decimal.ROUND_HALF_UP);
}

/**
 * 가성비. 100ml당 가격 1,000원 기준으로 정규화해 값이 너무 작아지지 않게 한다(Task 20).
 *
 * 높을수록 좋다. 가격이 0 이하이면 정의되지 않는다.
 * `src/sooljang/domain/metrics.py::value_for_money` 와 같은 공식이다.
 */
export function valueForMoney(
  rating: Decimal | null,
  pricePer100ml: Decimal | null,
): Decimal | null {
  if (rating === null || pricePer100ml === null || pricePer100ml.lessThanOrEqualTo(0)) {
    return null;
  }
  return quantizeRatio(rating.times(1000).dividedBy(pricePer100ml));
}

export const BottleState = {
  UNOPENED: "unopened",
  OPEN: "open",
  FINISHED: "finished",
  GIFTED: "gifted",
  SOLD: "sold",
} as const;

export const IN_STOCK_STATES: ReadonlySet<string> = new Set([
  BottleState.UNOPENED,
  BottleState.OPEN,
]);
export const DISPOSED_STATES: ReadonlySet<string> = new Set([BottleState.GIFTED, BottleState.SOLD]);

export interface PurchaseLot {
  quantity: number;
  volumeMl: number;
  unitListPrice: Decimal | null;
  unitPaidPrice: Decimal | null;
}

/** 구매 건 하나를 만든다. Python `PurchaseLot.__post_init__` 과 같은 검증을 한다. */
export function purchaseLot(
  quantity: number,
  volumeMl: number,
  unitListPrice: Decimal | null = null,
  unitPaidPrice: Decimal | null = null,
): PurchaseLot {
  if (quantity <= 0) throw new RangeError(`구매 병수는 양수여야 합니다: ${quantity}`);
  if (volumeMl <= 0) throw new RangeError(`용량은 양수여야 합니다: ${volumeMl}`);
  return { quantity, volumeMl, unitListPrice, unitPaidPrice };
}

function lotListTotal(lot: PurchaseLot): Decimal | null {
  return lot.unitListPrice === null ? null : lot.unitListPrice.times(lot.quantity);
}

function lotPaidTotal(lot: PurchaseLot): Decimal | null {
  return lot.unitPaidPrice === null ? null : lot.unitPaidPrice.times(lot.quantity);
}

function lotTotalVolumeMl(lot: PurchaseLot): number {
  return lot.volumeMl * lot.quantity;
}

export interface BottleRecord {
  status: string;
  openedOn: string | null;
  finishedOn: string | null;
}

export function bottleInStock(bottle: BottleRecord): boolean {
  return IN_STOCK_STATES.has(bottle.status);
}

export interface BottleTally {
  unopened: number;
  open: number;
  finished: number;
  gifted: number;
  sold: number;
}

const TALLY_STATUSES = ["unopened", "open", "finished", "gifted", "sold"] as const;

export function tallyBottles(bottles: Iterable<BottleRecord>): BottleTally {
  const tally: BottleTally = { unopened: 0, open: 0, finished: 0, gifted: 0, sold: 0 };
  for (const bottle of bottles) {
    if ((TALLY_STATUSES as readonly string[]).includes(bottle.status)) {
      tally[bottle.status as keyof BottleTally] += 1;
    }
  }
  return tally;
}

export function bottleTallyInStock(tally: BottleTally): number {
  return tally.unopened + tally.open;
}

export function bottleTallyDisposed(tally: BottleTally): number {
  return tally.gifted + tally.sold;
}

export function bottleTallyTotal(tally: BottleTally): number {
  return bottleTallyInStock(tally) + tally.finished + bottleTallyDisposed(tally);
}

export interface PriceMetrics {
  purchasedCount: number;
  pricedQuantity: number;
  paidQuantity: number;
  listTotal: Decimal | null;
  paidTotal: Decimal | null;
  avgListPrice: Decimal | null;
  avgPaidPrice: Decimal | null;
  pricePer100ml: Decimal | null;
  pricePer100mlPaid: Decimal | null;
  discountRate: Decimal | null;
  totalVolumeMl: number;
  /**
   * 통계 대시보드의 주종별·전체 집계가 제품 단위 가중 평균을 그룹 단위로 다시 계산할 때
   * 쓴다(`metrics_sql.py::product_price_metrics_query` 의 같은 이름 컬럼과 대응). 개별
   * 제품 지표 표시에는 쓰지 않는다.
   */
  listVolume: number;
  discountListTotal: Decimal | null;
  discountPaidTotal: Decimal | null;
}

export function priceMetricsHasPrices(metrics: PriceMetrics): boolean {
  return metrics.avgListPrice !== null || metrics.avgPaidPrice !== null;
}

interface WeightedTotal {
  amount: Decimal | null;
  quantity: number;
  volume: number;
}

/** 가격이 있는 구매 건만 합산한다. 금액이 있는 건이 없으면 `amount` 는 null 이다. */
function weightedTotal(lots: readonly PurchaseLot[], paid: boolean): WeightedTotal {
  let amount = new Decimal(0);
  let quantity = 0;
  let volume = 0;
  let found = false;
  for (const lot of lots) {
    const total = paid ? lotPaidTotal(lot) : lotListTotal(lot);
    if (total === null) continue;
    found = true;
    amount = amount.plus(total);
    quantity += lot.quantity;
    volume += lotTotalVolumeMl(lot);
  }
  return { amount: found ? amount : null, quantity, volume };
}

interface DiscountTotals {
  listAmount: Decimal | null;
  paidAmount: Decimal | null;
}

/** 정가와 실구매가가 모두 있는 구매 건만의 금액 합계. 할인율 분자·분모다. */
function discountTotalsOf(lots: readonly PurchaseLot[]): DiscountTotals {
  let listAmount = new Decimal(0);
  let paidAmount = new Decimal(0);
  let found = false;
  for (const lot of lots) {
    if (lot.unitListPrice === null || lot.unitPaidPrice === null) continue;
    found = true;
    listAmount = listAmount.plus(lot.unitListPrice.times(lot.quantity));
    paidAmount = paidAmount.plus(lot.unitPaidPrice.times(lot.quantity));
  }
  return found ? { listAmount, paidAmount } : { listAmount: null, paidAmount: null };
}

/** 할인율. 정가와 실구매가가 모두 있는 구매 건만으로 계산한다. */
function discountRateOf(totals: DiscountTotals): Decimal | null {
  if (totals.listAmount === null || totals.paidAmount === null || totals.listAmount.isZero()) {
    return null;
  }
  return quantizeRatio(new Decimal(1).minus(totals.paidAmount.dividedBy(totals.listAmount)));
}

export function computePriceMetrics(lots: readonly PurchaseLot[]): PriceMetrics {
  const purchasedCount = lots.reduce((sum, lot) => sum + lot.quantity, 0);
  const totalVolumeMl = lots.reduce((sum, lot) => sum + lotTotalVolumeMl(lot), 0);

  const list = weightedTotal(lots, false);
  const paid = weightedTotal(lots, true);

  const avgListPrice =
    list.amount !== null && list.quantity > 0
      ? quantizeMoney(list.amount.dividedBy(list.quantity))
      : null;
  const avgPaidPrice =
    paid.amount !== null && paid.quantity > 0
      ? quantizeMoney(paid.amount.dividedBy(paid.quantity))
      : null;
  // 여러 용량이 섞이면 가중 평균이어야 한다. 단일 용량이면 avg / (volume/100) 과 같다.
  const pricePer100ml =
    list.amount !== null && list.volume > 0
      ? quantizeMoney(list.amount.times(ML_PER_UNIT).dividedBy(list.volume))
      : null;
  const pricePer100mlPaid =
    paid.amount !== null && paid.volume > 0
      ? quantizeMoney(paid.amount.times(ML_PER_UNIT).dividedBy(paid.volume))
      : null;

  const discountTotals = discountTotalsOf(lots);

  return {
    purchasedCount,
    pricedQuantity: list.quantity,
    paidQuantity: paid.quantity,
    listTotal: list.amount,
    paidTotal: paid.amount,
    avgListPrice,
    avgPaidPrice,
    pricePer100ml,
    pricePer100mlPaid,
    discountRate: discountRateOf(discountTotals),
    totalVolumeMl,
    listVolume: list.volume,
    discountListTotal: discountTotals.listAmount,
    discountPaidTotal: discountTotals.paidAmount,
  };
}

export interface ProductMetrics {
  prices: PriceMetrics;
  bottles: BottleTally;
  inventoryValueAtCost: Decimal | null;
  averageDaysToFinish: Decimal | null;
  consistencyWarnings: string[];
}

/** ISO 날짜(`YYYY-MM-DD`) 두 값 사이의 일수. UTC 자정 기준으로 시간대 오차를 피한다. */
function daysBetween(openedOn: string, finishedOn: string): number {
  const start = Date.parse(`${openedOn}T00:00:00Z`);
  const end = Date.parse(`${finishedOn}T00:00:00Z`);
  return Math.round((end - start) / 86_400_000);
}

/**
 * 개봉→소진 평균 일수. 개봉일·소진일이 모두 있는 병만의 평균이다.
 *
 * 제품 하나(`computeProductMetrics`)뿐 아니라 컬렉션 전체(통계 요약)에도 그대로 쓴다 —
 * 병 목록을 평평하게 모아 넘기기만 하면 되므로 별도 컬렉션 버전을 만들 필요가 없다. 제품별
 * 평균을 다시 평균 내면(표본 크기를 무시하게 돼) 병이 적은 제품과 많은 제품이 똑같이
 * 반영되는 왜곡이 생긴다 — 항상 병 단위로 직접 평균한다.
 */
export function averageDaysToFinish(bottles: readonly BottleRecord[]): Decimal | null {
  const durations = bottles
    .filter((b): b is BottleRecord & { openedOn: string; finishedOn: string } =>
      Boolean(b.openedOn && b.finishedOn),
    )
    .map((b) => daysBetween(b.openedOn, b.finishedOn));
  if (durations.length === 0) return null;
  return quantizeRatio(
    durations.reduce((sum, d) => sum.plus(d), new Decimal(0)).dividedBy(durations.length),
  );
}

export function computeProductMetrics(
  lots: readonly PurchaseLot[],
  bottles: readonly BottleRecord[],
): ProductMetrics {
  const prices = computePriceMetrics(lots);
  const tally = tallyBottles(bottles);

  const warnings: string[] = [];
  const bottleCount = bottleTallyTotal(tally);
  if (bottleCount !== prices.purchasedCount) {
    warnings.push(`병 레코드 수(${bottleCount})가 구매 병수(${prices.purchasedCount})와 다릅니다`);
  }

  const inventoryValueAtCost =
    prices.avgPaidPrice !== null
      ? quantizeMoney(prices.avgPaidPrice.times(bottleTallyInStock(tally)))
      : null;

  return {
    prices,
    bottles: tally,
    inventoryValueAtCost,
    averageDaysToFinish: averageDaysToFinish(bottles),
    consistencyWarnings: warnings,
  };
}
